packages/server/src/middleware/cleanup.ts

```diff
@@ -0,0 +1,30 @@
+import { Ctx } from "@budibase/types"
+import { context } from "@budibase/backend-core"
+
+export default async (ctx: Ctx, next: any) => {
+  const resp = await next()
+
+  const current = context.getCurrentContext()
+  if (!current || !current.cleanup) {
+    return resp
+  }
+
+  let errors = []
+  for (let fn of current.cleanup) {
+    try {
+      await fn()
+    } catch (e) {
+      // We catch errors here to ensure we at least attempt to run all cleanup
+      // functions. We'll throw the first error we encounter after all cleanup
+      // functions have been run.
+      errors.push(e)
+    }
+  }
+  delete current.cleanup
+
+  if (errors.length > 0) {
+    throw errors[0]
```

Comment:
Is this safe - do we run the risk of hiding some issues if there is multiple errors?

Reply:
If multiple "finalisers" failed, yes, we'd hide some errors. I considered making some sort of `CompositeError` class but then you'd be hiding the stack traces of all errors. Seemed like the lesser of two evils at the time.

Reply:
Should we at least log the errors that we can then see that something else happened?

Reply:
I've wrapped each cleanup function in an APM span. This will make them show up in APM tracing, and any error thrown from them will get attached to the span and then show up in DataDog's error tracking dashboard.